File: 2375_Construct_Smallest_Number_From_DI_String/solution.py
class Solution:
    def smallestNumber(self, pattern: str) -> str:
        n = len(pattern)
        result = list(range(1, n + 2))

        i = 0
        while i < n:
            if pattern[i] == "D":
                start = i
                while i < n and pattern[i] == "D":
                    i += 1
                end = i
                result[start : end + 1] = reversed(result[start : end + 1])
            else:
                i += 1

        return "".join(str(x) for x in result)

File: 2375_Construct_Smallest_Number_From_DI_String/solution.c
char* reverse(char* result, int start, int end) {
    while (start < end) {
        char temp = result[start];
        result[start] = result[end];
        result[end] = temp;
        start++;
        end--;
    }
    return result;
}
char* smallestNumber(char* pattern) {
    int n = strlen(pattern);
    char* result = malloc(n + 2);

    for (int i = 0; i <= n; i++) result[i] = '0' + (i + 1);
    result[n + 1] = '\0';

    for (int i = 0; i < n; i++) {
        if (pattern[i] == 'D') {
            int start = i;
            while (i < n && pattern[i] == 'D') i++;
            int end = i;
            reverse(result, start, end);
        }
    }
    return result;
}
